### ISSUE TITLE
Gas usage benchmarks for bridge contracts

### ISSUE THREAD

**Issue by akolotov** [poanetwork/poa-bridge-contracts#47]
Since it is critical to have clear understanding how much gas consumed by the bridge contracts in different scenarios and different deployments it is necessary to perform the corresponding measurements.

Goals are the following:
1. To prepare a sheet in the bridge documentation with description of gas consumed on Home and Foreign side in different scenarios.
2. To identify potential places in the code where gas consumption could be optimized.

**Comment by patitonar** [comment]
@akolotov As a first step to get knowledge of gas consumed by bridge contracts I think we can add https://github.com/cgewecke/eth-gas-reporter which is a library to generate gas usage reports of methods and deployments, based on our unit tests. What do you think?

**Comment by akolotov** [comment]
Looks interesting. Could you estimate how much effort it will take to integrate this? I would avoid spending much time on this issue. From my perspective it is one-time-run every time we add new bridge mode and it will not take more than 2 days.

**Comment by patitonar** [comment]
Installation is pretty straightforward, it took me 10 minutes. If you are OK I can open a PR for this.
Here is an example of the output for `refactor_v1` branch.

```
·------------------------------------------------------------------------------------------|-----------------------------------·
|                                           Gas                                            ·  Block limit: 17592186044415 gas  │
·······················································|···································|····································
|  Methods                                             ·            1 gwei/gas             ·          205.39 usd/eth           │
·····························|·························|···········|···········|···········|··················|·················
|  Contract                  ·  Method                 ·  Min      ·  Max      ·  Avg      ·  # calls         ·  usd (avg)     │
·····························|·························|···········|···········|···········|··················|·················
|  BridgeValidators          ·  addValidator           ·        -  ·        -  ·    53178  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  BridgeValidators          ·  initialize             ·   187738  ·   218775  ·   212568  ·               5  ·          0.04  │
·····························|·························|···········|···········|···········|··················|·················
|  BridgeValidators          ·  removeValidator        ·        -  ·        -  ·    24400  ·               3  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  BridgeValidators          ·  setRequiredSignatures  ·        -  ·        -  ·    29790  ·               9  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  burn                   ·        -  ·        -  ·    18238  ·               1  ·          0.00  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  claimTokens            ·        -  ·        -  ·    44097  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  mint                   ·    53347  ·    68475  ·    66400  ·              16  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  transfer               ·    37649  ·    86544  ·    47457  ·               8  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  transferAndCall        ·    58390  ·   166184  ·    93845  ·               6  ·          0.02  │
·····························|·························|···········|···········|···········|··················|·················
|  ERC677BridgeToken         ·  transferOwnership      ·        -  ·        -  ·    30880  ·              10  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  EternalStorageProxy       ·  upgradeTo              ·    35871  ·    65871  ·    55871  ·               6  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  EternalStorageProxy       ·  upgradeToAndCall       ·   178273  ·   262801  ·   235720  ·               5  ·          0.05  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeErcToErc     ·  claimTokens            ·        -  ·        -  ·    52115  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeErcToErc     ·  executeSignatures      ·    73759  ·   102003  ·    88408  ·               5  ·          0.02  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeErcToErc     ·  initialize             ·        -  ·        -  ·   129357  ·               2  ·          0.03  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  claimTokens            ·        -  ·        -  ·    47522  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  claimTokensFromErc677  ·        -  ·        -  ·    50742  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  executeSignatures      ·    89179  ·   132423  ·   115853  ·               5  ·          0.02  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  initialize             ·        -  ·        -  ·   213557  ·               7  ·          0.04  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  setMaxPerTx            ·        -  ·        -  ·    32249  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc  ·  setMinPerTx            ·        -  ·        -  ·    32774  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc        ·  executeAffirmation     ·    79336  ·   134585  ·   109421  ·               5  ·          0.02  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc        ·  initialize             ·   212299  ·   213195  ·   212896  ·               3  ·          0.04  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc        ·  setMaxPerTx            ·        -  ·        -  ·    32084  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc        ·  setMinPerTx            ·        -  ·        -  ·    32673  ·               1  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc        ·  submitSignature        ·   166504  ·   275159  ·   228620  ·               7  ·          0.05  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  executeAffirmation     ·    79270  ·   107669  ·    87580  ·               7  ·          0.02  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  initialize             ·   190051  ·   190947  ·   190648  ·               3  ·          0.04  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  setDailyLimit          ·        -  ·        -  ·    32367  ·               3  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  setMaxPerTx            ·        -  ·        -  ·    32040  ·               3  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  setMinPerTx            ·        -  ·        -  ·    32651  ·               2  ·          0.01  │
·····························|·························|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc     ·  submitSignature        ·   166607  ·   275135  ·   228623  ·               7  ·          0.05  │
·····························|·························|···········|···········|···········|··················|·················
|  Deployments                                         ·                                   ·  % of limit      ·                │
·······················································|···········|···········|···········|··················|·················
|  BridgeValidators                                    ·        -  ·        -  ·  1144207  ·             0 %  ·          0.24  │
·······················································|···········|···········|···········|··················|·················
|  ERC677BridgeToken                                   ·  1416059  ·  1417083  ·  1416707  ·             0 %  ·          0.29  │
·······················································|···········|···········|···········|··················|·················
|  EternalStorageProxy                                 ·        -  ·        -  ·   378510  ·             0 %  ·          0.08  │
·······················································|···········|···········|···········|··················|·················
|  ForeignBridgeErcToErc                               ·        -  ·        -  ·  2480305  ·             0 %  ·          0.51  │
·······················································|···········|···········|···········|··················|·················
|  ForeignBridgeNativeToErc                            ·        -  ·        -  ·  2767086  ·             0 %  ·          0.57  │
·······················································|···········|···········|···········|··················|·················
|  HomeBridgeErcToErc                                  ·        -  ·        -  ·  3526762  ·             0 %  ·          0.72  │
·······················································|···········|···········|···········|··················|·················
|  HomeBridgeNativeToErc                               ·        -  ·        -  ·  3361663  ·             0 %  ·          0.69  │
·------------------------------------------------------|-----------|-----------|-----------|------------------|----------------·

```

**Comment by akolotov** [comment]
Yes, please create the PR. I would also request to change our tests as so 3 or more validators were used (with the consensus `2 of 3` or `3 of 5`). It will allow to get more practical information about average gas usage.

**Comment by akolotov** [comment]
Addressed in #99
